### PR TITLE
Refactor Setting - Service and provide service configuration view

### PR DESCRIPTION
**Pull request by CanglongCl**

### CODE REVIEW

**Comment by CanglongCl** [comment]
#324 
For service configuration, only a framework has been written, and the details of the configuration need to be filled in.

**Comment by phlpsong** [comment]
I tried this change, and it looks good!

Meanwhile, I've noticed some minor UI issues:

1. The Toggle size increased.
2. The behavior of the service cell highlights has changed; could you double-check it? (I saw that the list style changes, I'm not sure if we can retain the old behavior.)
3. The `Picker` cannot be fully shown in the sidebar.
4. The `Text` for no selection and no configuration should have a `secondary` color for a better appearance.

**Comment by tisfeng** [comment]
Yes, I'd like to see the UI of the service list to be as consistent as possible with the previous one if there are no special requirements, especially the display style of the Picker and cell, which are not as recognizable in comparison to the new UI as they were before.

image

image

**Comment by tisfeng** [comment]
The rewritten `ServiceTab.swift` looks cleaner and more structured, we can refer to this code when we write SwiftUI later.

**Comment by CanglongCl** [comment]
> I tried this change, and it looks good!
> 
> Meanwhile, I've noticed some minor UI issues:
> 
> 1. The Toggle size increased.
> 2. The behavior of the service cell highlights has changed; could you double-check it? (I saw that the list style changes, I'm not sure if we can retain the old behavior.)
> 3. The `Picker` cannot be fully shown in the sidebar while using English.
> 4. The `Text` for no selection and no configuration should have a `secondary` color for a better appearance.

For point 4, I don't this text style should be secondary. Usually, secondary mean disabled. It may confuse user. 

@tisfeng How do you think?

**Comment by phlpsong** [comment]
For point 4 I think you are right.  Let's keep the current setting.

**Comment by tisfeng** [comment]
> The Text for no selection and no configuration should have a secondary color for a better appearance.

Currently, it looks good to me, do you have any good idea for this? @phlpsong 

目前这个效果对我来说看起来不错，对于这个，你有更好的想法吗？

**Comment by phlpsong** [comment]
> > The Text for no selection and no configuration should have a secondary color for a better appearance.
> 
> Currently, it looks good to me, do you have any good idea for this? @phlpsong
> 
> 目前这个效果对我来说看起来不错，对于这个，你有更好的想法吗？

No other issues, LGTM

**Comment by tisfeng** [comment]
ok.

**Comment by tisfeng** [comment]
I'll add the rest of the service configuration later in the evening when I have time.

**Comment by CanglongCl** [comment]
This feature is almost completed, it's recommended to approve & merge before start add other configurations.

**Comment by tisfeng** [comment]
I tested this code with no problems.

The only issue is that switching window types multiple times may cause lag, but this seems to be an internal issue with the `allServices:` method, which I'll troubleshoot later.

**Comment by tisfeng** [comment]
@CanglongCl  This is a strange issue, I have tested it and found that when the program is first running, switching Tabs doesn't cause the problem, it seems that after running for a while, switching Tabs causes `allServices:` to be called multiple times, and the CPU usage is very high, this is an issue you may want to look at.

https://github.com/tisfeng/Easydict/assets/25194972/808ed75a-b893-4f7b-a331-0ef02da0f2f9

**Comment by CanglongCl** [comment]
> @CanglongCl This is a strange issue, I have tested it and found that when the program is first running, switching Tabs doesn't cause the problem, it seems that after running for a while, switching Tabs causes `allServices:` to be called multiple times, and the CPU usage is very high.
> 
>  service-issue.mp4

I didn't reproduce the issue. But I did a small refactor so now only the windowType changes will trigger calling to `allServices`

**Comment by tisfeng** [comment]
@CanglongCl  The new code is much improved, but it still calls `allServices:` multiple times and sometimes lags.

I found the stable way to reproduce it: after launching the app, first show the mini window, then open the Settings-Services page, and then the multiple calls will occur.

https://github.com/tisfeng/Easydict/assets/25194972/ebd3f28e-2c8a-41e2-9692-91b41886c73d

**Comment by CanglongCl** [comment]
@tisfeng Thank you. I've reproduced the issue. I identify multi-times `allServices` is called in `EZBaseQueryViewController` instead of SwiftUI View. I'm not familiar with objc and UIKit. Could you dive into it? 

image

**Comment by tisfeng** [comment]
I have fixed this issue.

The problem is that every time the `ServiceItemView` is initialized, the set `enabled` method is called to post an update notification, and the current service page has so many `ServiceItemView`.

image

**Comment by CanglongCl** [comment]
> I have fixed this issue.
> 
> The problem is that every time the `ServiceItemView` is initialized, the set `enabled` method is called to post an update notification, and the current service page has so many `ServiceItemView`.
> 
> image

Thank you! It is really a weird behavior of Toggle in SwiftUI...

**Comment by tisfeng** [comment]
Last question, the list on the left side of the service page seems to have a weird double animation, could you take a look? @CanglongCl 


https://github.com/tisfeng/Easydict/assets/25194972/7897e2b6-fa53-4ab8-be29-85baddab0fe7

**Comment by CanglongCl** [comment]
> Last question, the list on the left side of the service page seems to have a weird double animation, could you take a look? @CanglongCl
> 
>  double-animation.mp4

Fixed

**Comment by CanglongCl** [comment]
> > I have fixed this issue.
> > The problem is that every time the `ServiceItemView` is initialized, the set `enabled` method is called to post an update notification, and the current service page has so many `ServiceItemView`.
> > image
> 
> Thank you! It is really a weird behavior of Toggle in SwiftUI...

I figured out it. It is because I called improperly set `enabled` while initializing the wrapper.

**Comment by tisfeng** [comment]
I tested the new code, it looks good to me 👍

**Review by phlpsong — DISMISSED** [review]
LGTM

File: Easydict/NewApp/View/SettingView/Tabs/ServiceTab.swift [hunk 1]
```diff
@@ -6,141 +6,209 @@
...
-                .tag(EZWindowType.fixed.rawValue)
-
-            Text("main_window")
-                .tag(EZWindowType.main.rawValue)
-        }
-        .padding()
-        .pickerStyle(.segmented)
-        .onChange(of: windowTypeValue) { type in
-            loadService(type: type)
-            selectedIndex = nil
```
[the remark this answers is not in the document]
Comment (tisfeng) [reply Easydict/NewApp/View/SettingView/Tabs/ServiceTab.swift:35]: Yes, I didn't notice this.

**Comment by tisfeng** [comment]
Please review the code, if no problem, I will merge this PR tonight.

**Comment by phlpsong** [comment]
I tested the latest commit, looks good to me 🎉

**Review by phlpsong — APPROVED** [review]

**Comment by Jerry23011** [comment]
The new UI is awesome, super excited to see it in production 🥳🙌

Just some notes on minor issues here:
- In settings, the button appears to experience lag when transitioning from other tabs to General
- In Services tab, the list seems to wear off along with the elements when scrolling to the top/bottom
截屏2024-01-16 19 14 20
截屏2024-01-16 19 14 35

- Scrolling in General tab feels laggy on 120Hz display (may be a SwiftUI performance issue 🥲)
- Settings window often fails to appear in the front (again, likely to be on SwiftUI's side 🥲🥲)

https://github.com/tisfeng/Easydict/assets/89069957/8999d965-d0f6-42ca-8c73-5825aa992d22


I suggest keeping these features on beta for a while before Apple fixes them or

**Review by Jerry23011 — APPROVED** [review]

**Comment by CanglongCl** [comment]
> The new UI is awesome, super excited to see it in production 🥳🙌
> 
> Just some notes on minor issues here:
> 
> * In settings, the button appears to experience lag when transitioning from other tabs to General
> * In Services tab, the list seems to wear off along with the elements when scrolling to the top/bottom
> 
> 截屏2024-01-16 19 14 20 截屏2024-01-16 19 14 35
> * Scrolling in General tab feels laggy on 120Hz display (may be a SwiftUI performance issue 🥲)
> * Settings window often fails to appear in the front (again, likely to be on SwiftUI's side 🥲🥲)
> 
>  2024-01-16.19.24.03-1.mov 
> I suggest keeping these features on beta for a while, either until Apple fixes them or until we find workarounds.



> The new UI is awesome, super excited to see it in production 🥳🙌
> 
> Just some notes on minor issues here:
> 
> * In settings, the button appears to experience lag when transitioning from other tabs to General
> * In Services tab, the list seems to wear off along with the elements when scrolling to the top/bottom
> 
> 截屏2024-01-16 19 14 20 截屏2024-01-16 19 14 35
> * Scrolling in General tab feels laggy on 120Hz display (may be a SwiftUI performance issue 🥲)
> * Settings window often fails to appear in the front (again, likely to be on SwiftUI's side 🥲🥲)
> 
>  2024-01-16.19.24.03-1.mov 
> I suggest keeping these features on beta for a while, either until Apple fixes them or until we find workarounds.

For point 1 & 4, I think it SwiftUI's issue. 

For point 2, I think it is the default behavior of `List` view.  It indicates the end of scrolling. If it's a big problem, I will take more time to fix it. @tisfeng 

For point 4, I think we can find a way to fix it, but I'm not familiar to window management in SwiftUI. Maybe we should handle it in another PR.

**Comment by phlpsong** [comment]
For point 4 I think it's a SwiftUI issue, same problem here [https://developer.apple.com/forums/thread/731628](https://developer.apple.com/forums/thread/731628).

**Comment by tisfeng** [comment]
@Jerry23011 You're very observant. 👍

For points 1 and 3, there does seem to be a slight issue, but it's not very noticeable perceptually, probably because there are more elements in the General Tab, not sure if it's a performance issue with SwiftUI.

For points 2 and 4, release mode performs much better compared to debug mode, seems strange.

For this kind of problem, which is not very serious, we can check it a little bit, and if it's a problem with SwiftUI itself, we can ignore it for now and leave it to Apple to fix it.

**Comment by tisfeng** [comment]
@Jerry23011 You can open a new issue to record these problem, we can handle them in another PR.

If there are no more questions, I will merge this PR later.

**Pull request merged**